### ISSUE TITLE
RStudio crashes durin this excersize

### ISSUE THREAD

**Issue by glensbo** [rstudio/tensorflow#432]
Following the code:

### DL-related packages
library(tensorflow)
library(keras)
library(tfdatasets)
library(tfautograph)
library(reticulate)

### going to need those later
library(tidyverse)
library(cowplot)
geyser <- download.file(
    "https://raw.githubusercontent.com/rstudio/ai-blog/master/docs/posts/2020-07-20-fnn-lstm/data/geyser.csv",
    "data/geyser.csv")
geyser <- read_csv("https://raw.githubusercontent.com/rstudio/ai-blog/master/docs/posts/2020-07-20-fnn-lstm/data/geyser.csv", col_names = FALSE) %>% select(X1) %>% pull() %>% unclass()
geyser <- scale(geyser)
# varies per dataset; see below 
n_timesteps <- 60
batch_size <- 32

### transform into [batch_size, timesteps, features] format required by RNNs
gen_timesteps <- function(x, n_timesteps) {
    do.call(rbind,
            purrr::map(seq_along(x),
                       function(i) {
                           start <- i
                           end <- i + n_timesteps - 1
                           out <- x[start:end]
                           out
                       })
    ) %>%
        na.omit()
}
n <- 10000
train <- gen_timesteps(geyser[1:(n/2)], 2 * n_timesteps)
test <- gen_timesteps(geyser[(n/2):n], 2 * n_timesteps) 

dim(train) <- c(dim(train), 1)
dim(test) <- c(dim(test), 1)
### split into input and target  
x_train <- train[ , 1:n_timesteps, , drop = FALSE]
y_train <- train[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]

x_test <- test[ , 1:n_timesteps, , drop = FALSE]
y_test <- test[ , (n_timesteps + 1):(2*n_timesteps), , drop = FALSE]

### create tfdatasets
ds_train <- tensor_slices_dataset(list(x_train, y_train)) %>%
    dataset_shuffle(nrow(x_train)) %>%
    dataset_batch(batch_size)

ds_test <- tensor_slices_dataset(list(x_test, y_test)) %>%
    dataset_batch(nrow(x_test))

###When I run the code below: # create tfdatasets RStudio crashes
further info: https://blogs.rstudio.com/ai/posts/2020-07-20-fnn-lstm/

**Comment by skeydan** [comment]
what is the output you get for 

```
tensorflow::tf_config()
```

?

**Comment by glensbo** [comment]
Thanks for responding - RStudio crashes!

**Comment by glensbo** [comment]
Made a brand new project in Rstudio and ran the code:
> install_tensorflow()
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done
# All requested packages already installed.
Requirement already up-to-date: tensorflow==2.2.0 in /Users/

AND THEN THESE MESSAGES CAME FORWARD:
d<2.3.0,>=2.2.0->tensorflow==2.2.0) (0.4.8)

Installation complete.
Restarting R session...
> tensorflow::tf_config()
Installation of TensorFlow not found. ???????
Python environments searched for 'tensorflow' package:
 /Users/henrikglensbo/Library/r-miniconda/envs/r-reticulate/bin/python3.6
You can install TensorFlow using the install_tensorflow() function.

**Comment by glensbo** [comment]
Then I followed the installation guide: https://tensorflow.rstudio.com/installation/
Running this code:
library(tensorflow)
tf$constant("Hellow Tensorflow")
Rstudio crashes

**Comment by glensbo** [comment]
Seems as if there is a problem with - as you stated - the installation of tensorflow - I keep running around in loops:
Installation complete.


Restarting R session...

> detach("package:tensorflow", unload = TRUE)
> library(tensorflow)
> tf$constant("Hellow Tensorflow")
Error: Installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /Users/henrikglensbo/Library/r-miniconda/envs/r-reticulate/bin/python3.6

You can install TensorFlow using the install_tensorflow() function.

Restarting R session...

> install.packages("tensorflow")
trying URL 'https://cran.rstudio.com/bin/macosx/el-capitan/contrib/3.6/tensorflow_2.2.0.tgz'
Content type 'application/x-gzip' length 142881 bytes (139 KB)
==================================================
downloaded 139 KB


The downloaded binary packages are in
	/var/folders/pf/cbcn8vs50zb7vzmxblqqpxym0000gn/T//Rtmp9JGaKe/downloaded_packages
> library(tensorflow)
> tf$constant("Hellow Tensorflow")
Error: Installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /Users/henrikglensbo/Library/r-miniconda/envs/r-reticulate/bin/python3.6

You can install TensorFlow using the install_tensorflow() function.

**Comment by glensbo** [comment]
Then I arrived at this: > hello <- tf$constant("Hello")
2020-07-24 09:07:26.842007: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-07-24 09:07:26.913808: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fe1a992d780 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-24 09:07:26.913826: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
> print(hello)
tf.Tensor(b'Hello', shape=(), dtype=string)

I will try to install an earlier version of tensorflow

**Comment by glensbo** [comment]
Now I deinstalled tensorflow 2.2.0 and try installing 2.1.0 result:
Installation complete.


Restarting R session...

> library(tensorflow)
Error in library(tensorflow) : there is no package called ‘tensorflow’
> tf$constant("Hellow Tensorflow")
Error: object 'tf' not found
> install_tensorflow(version = "2.1.0")
Error in install_tensorflow(version = "2.1.0") : 
  could not find function "install_tensorflow"
> library(tensorflow)
Error in library(tensorflow) : there is no package called ‘tensorflow’
> install_tensorflow(version = "2.1.0")
Error in install_tensorflow(version = "2.1.0") : 
  could not find function "install_tensorflow"

**Comment by glensbo** [comment]
I think tensorflow searches in the wrong place: Error: Installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /Users/henrikglensbo/Library/r-miniconda/envs/r-reticulate/bin/python3.6

from previous post: The downloaded binary packages are in
/var/folders/pf/cbcn8vs50zb7vzmxblqqpxym0000gn/T//Rtmp9JGaKe/downloaded_packages

**Comment by glensbo** [comment]
https://github.com/rstudio/tensorflow/issues/144 - installing Rstudio/reticulate....
RESULT:
Installation complete.
Restarting R session...
> library(tensorflow)
> tf$constant("Hellow Tensorflow")
2020-07-24 16:52:44.395530: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-07-24 16:52:44.421525: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7f889d994260 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-07-24 16:52:44.421542: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
tf.Tensor(b'Hellow Tensorflow', shape=(), dtype=string)

**Comment by skeydan** [comment]
Hi, do you mean you got this fixed? :-)

**Comment by glensbo** [comment]
No - by I am trying my best to search for solutions - and testing options - I hope to get thing solved but I am unsure for the time being of how to move forward - as whatever I do - sends me around in loops. I will let you know if I find a solution. Thanks for helping me along.

Regards Henrik

**Comment by skeydan** [comment]
Sorry to hear this causes you so much work!!

However, what you posted in the last comment looked like it worked?

This...

```
Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
````

etc. is just a warning...

And you seem to get the desired output:

```
tf.Tensor(b'Hellow Tensorflow', shape=(), dtype=string)
```

**Comment by glensbo** [comment]
Yes but when I run the programme listed at the top of the discussion it crashes RStudio again - Once I was able to get another message trying to run tensorflow::tf_config(): OMP: Error #15: Initializing libomp.dylib, but found libomp.dylib already initialized.
OMP: Hint: This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results.
The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library.
As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results.

Which I will dig into again tomorrow - thanks for your support

**Comment by glensbo** [comment]
# DL-related packages
> library(tensorflow)
> tensorflow::tf_config()
TensorFlow v2.2.0 ()
Python v3.6 (~/Library/r-miniconda/envs/r-reticulate/bin/python)
> library(tfautograph)
> library(reticulate)
> library(keras)

I have done al sorts of changes but have now identified "tfdatasets" as the reason for the crash

**Comment by glensbo** [comment]
Got a new error when running -> 
> install.packages("keras", type = "source")
and then
> library(keras)
> install_keras()ERROR: tensorflow 2.2.0 has requirement scipy==1.4.1; python_version >= "3", but you'll have scipy 1.5.2 which is incompatible.
Installing collected packages: scipy, pyyaml, keras, tensorflow-hub, Pillow
  Attempting uninstall: scipy
    Found existing installation: scipy 1.4.1
    Uninstalling scipy-1.4.1:
      Successfully uninstalled scipy-1.4.1
Successfully installed Pillow-7.2.0 keras-2.4.3 pyyaml-3.12 scipy-1.5.2 tensorflow-hub-0.8.0

Installation complete.

**Comment by skeydan** [comment]
For that last error, can you try using the development version of Keras?

```
devtools::install_github("rstudio/keras")
```

BTW `install_keras` is seldom needed (in any case, not for that blog post). Nearly all Keras functionality is included in TensorFlow.

**Comment by glensbo** [comment]
###It gives this result (when I update all packages):
Error in parse(outFile) : 
  /private/var/folders/pf/cbcn8vs50zb7vzmxblqqpxym0000gn/T/Rtmp023b0p/R.INSTALL30ef3effc9cf/reticulate/R/zzz.R:62:0: unexpected end of input
60:     py_finalize()
61: }
   ^
ERROR: unable to collate and parse R files for package ‘reticulate’
* removing ‘/Library/Frameworks/R.framework/Versions/3.6/Resources/library/reticulate’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.6/Resources/library/reticulate’
Error: Failed to install 'keras' from GitHub:
  Failed to install 'reticulate' from GitHub:
  (converted from warning) installation of package ‘/var/folders/pf/cbcn8vs50zb7vzmxblqqpxym0000gn/T//RtmpneHaAW/file2fbf1ee79d6a/reticulate_1.16-9000.tar.gz’ had non-zero exit status

### It succeeds when I update only cran packages
* DONE (keras)

**Comment by glensbo** [comment]
tfdatasets still causes the code to crash RStudio

**Comment by glensbo** [comment]
Looking into tfdatasets:

library(tfdatasets)

# create specification for parsing records from an example file
iris_spec <- csv_record_spec("~/Documents/RFolder/BOGEN/Bogen_Kapt4/Data/Normal/NormalSl1.csv") 
### This line crashes RStudio
dataset <- text_line_dataset("~/Documents/RFolder/BOGEN/Bogen_Kapt4/Data/Normal/NormalSl1.csv", record_spec = iris_spec)

**Comment by glensbo** [comment]
This is weird:
devtools::install_github("rstudio/tfdatasets")
Downloading GitHub repo rstudio/tfdatasets@HEAD
These packages have more recent versions available.
It is recommended to update all of them.
Which would you like to update?

1: All                                               
2: CRAN packages only                                
3: None                                              
4: reticulate (6e44eff5f... -> 0b6d19801...) [GitHub]

Enter one or more numbers, or an empty line to skip updates:2
✓  checking for file ‘/private/var/folders/pf/cbcn8vs50zb7vzmxblqqpxym0000gn/T/Rtmp8QHyfl/remotes331874a2290/rstudio-tfdatasets-e4864e4/DESCRIPTION’ (376ms)
─  preparing ‘tfdatasets’:
✓  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Removed empty directory ‘tfdatasets/.github/workflows’
   Removed empty directory ‘tfdatasets/.github’
   Removed empty directory ‘tfdatasets/dev’
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘tfdatasets_2.2.0.9000.tar.gz’
   
* installing *source* package ‘tfdatasets’ ...
** using staged installation
** R
** data
*** moving datasets to lazyload DB
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (tfdatasets)
> library(tfdtasets)
Error in library(tfdtasets) : there is no package called ‘tfdtasets’
> install_tensorflow()
Error in install_tensorflow() : 
  could not find function "install_tensorflow"
> Can anybody help explaining how I get on from here - which commands are necessary?

**Comment by skeydan** [comment]
```
# you have a typo here :-)
library(tfdtasets)
Error in library(tfdtasets) : there is no package called ‘tfdtasets’

# and here, tensorflow is not loaded
install_tensorflow()
Error in install_tensorflow() :
could not find function "install_tensorflow"

```

**Comment by glensbo** [comment]
Thanks a lot

**Comment by glensbo** [comment]
Finally I found this:
https://github.com/rstudio/tensorflow/commit/4e1e11d6ba2fe7efe1a03356f96172dbf8db365e
and managed to get Tensorflow working by pointing to a specific python path in my Mac Os 10.14.06:
Sys.setenv(TENSORFLOW_PYTHON="/Users/_MyName_/opt/anaconda3/bin/") 
tensorflow::install_tensorflow()

**Comment by glensbo** [comment]
Running smoothly !!!! Just drawing your attention to:
WARNING:tensorflow:Layer gaussian_noise is casting an input tensor from dtype float64 to the layer's dtype of float32, which is new behavior in TensorFlow 2.  The layer has dtype float32 because it's dtype defaults to floatx.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

Warning message:
The `x` argument of `as_tibble.matrix()` must have unique column names if `.name_repair` is omitted as of tibble 2.0.0.
Using compatibility `.name_repair`.
This warning is displayed once every 8 hours.
Call `lifecycle::last_warnings()` to see where this warning was generated.

**Comment by skeydan** [comment]
Wonderful - happy to hear that!!

The above are just warnings and can be ignored :-)

I'll close the issue then, happy you got it working.
